### PR TITLE
Documentation for TRIM, RTRIM and LTRIM OQL functions

### PR DESCRIPTION
**Pull request by passalaqua**
Requires team review. To be published with Studio Pro 11.11

### CODE REVIEW

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 1]
```diff
@@ -1281,6 +1284,41 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
 This query can no longer take advantage of an index for `LastName` for comparison, resulting in a performance decrease.
 {{% /alert %}}
 
+### LTRIM{#ltrim}
+
+Trims characters to the left of a `string`.
```
Comment (basdebakker) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: I think the phrasing could be improved. There are no characters to the left of the string. "Trims characters from the left of"? Maybe the documentation team has a better suggestion?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 2]
```diff
@@ -1281,6 +1284,41 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
...
+#### Examples
+
+```sql
+SELECT LTRIM(LastName, 'e') FROM Sales.Order WHERE Price = 1.50000001
+```
+
+| LastName |
+|:---------|
+| Do       |
+| Moos     |
```
Comment (basdebakker) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: I think the LTRIM and RTRIM examples are mixed up.

Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: It's trimming my right, not your right!

(Fixed.)

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 1 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: I used leading and trailing, as that is also what most of our DBs seem to favor. How about now?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 3]
```diff
@@ -1559,6 +1632,41 @@ ORDER BY LastName LIMIT 1
 |:-------------|:---------------|:----------------|:-----------------|
 | logical      | logic          | logical         | *(empty string)* |
 
+### TRIM{#ltrim}
+
+Removes leading and trailing characters from a `string`. If no `character` is specified for trimming, whitespace is used.
```
Comment (basdebakker) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: In my vocabulary, "whitespace" would include tabs, line feeds, etc. But it only removes space characters.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 1 above)
Comment (basdebakker) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: Yes, that is better.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 4]
```diff
@@ -1559,6 +1632,41 @@ ORDER BY LastName LIMIT 1
 |:-------------|:---------------|:----------------|:-----------------|
 | logical      | logic          | logical         | *(empty string)* |
 
+### TRIM{#ltrim}
```
Comment (basdebakker) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: The anchor is incorrect.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 3 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: I also used `space character` in the description later. Fixed all three cases.

**Comment by basdebakker** [comment]
Approved by datastorage team.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 5]
```diff
@@ -1329,6 +1332,41 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
...
+
+```sql
+SELECT LTRIM(LastName, 'D') FROM Sales.Order WHERE Price = 1.50000001
+```
+
+| LastName |
+|:---------|
+| oe       |
+| Moose    |
+
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1375]: What happens if the trim character is the only character of the string? Do you get NULL or a zero-length string?

`LTRIM(String, 'D')` `DDDDD`

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 6]
```diff
@@ -1329,6 +1332,41 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
 This query can no longer take advantage of an index for `LastName` for comparison, resulting in a performance decrease.
 {{% /alert %}}
 
+### LTRIM{#ltrim}
+
+Removes leading characters from a `string`. If no `character` is specified for trimming, space is used.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: Does the language of the app make any difference. For example, if the string is Arabic, LTRIM will still trim from the left (rather than the leading characters on the right).

**Review by MarkvanMents — COMMENTED** [review]
Hi @passalaqua 
I've made a couple of changes in https://github.com/mendix/docs/pull/11058/changes/305c9c93ef7478afc35369af33350c0d75d683c1..c69e85379dc8f88fe5dc98fa1a4fa222e8aab018 and have a couple of questions.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 5 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1375]: You should get a zero-length string.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 6 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: As far as I could see, `ltrim` and `rtrim` remove leading and trailing bytes from the sequence of bytes of the string with no consideration as to how it is rendered.

If the sequence of bytes includes bidi characters (like unicode's LRI/RLI), then it might interfere with the `ltrim` behavior as the string does not actually start with spaces (or whichever other character to be trimmed).

Ultimately, we leave it up to the database as to how to handle these cases.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 7]
```diff
@@ -1329,6 +1332,49 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
...
+##### character
+
+`character` is an optional single character string expression containing the character to remove from the start of the string. If omitted, the space character is used instead.
+
+{{% alert color="info" %}}
+Only a single character is supported. `character` parameters with more than one character may not work in all supported databases.
+{{% /alert %}}
+
+{{% alert color="info" %}}
+Like with other String functions, case sensitivity of the `LOCATE` function depends on the database. See [Behavior of Case Sensitivity by Database Type](/refguide/case-sensitive-database-behavior/#behavior-of-case-sensitivity-by-database-type) for details.
```
Comment (passalaqua) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: Why is this alert box here?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 8]
```diff
@@ -1507,6 +1553,49 @@ SELECT ROUND((Price : 7), 2) as RoundedPrice, Price : 7 FROM Sales.Order
...
+##### character
+
+`character` is an optional single character string expression containing the character to remove from the end of the string. If omitted, the space character is used instead.
+
+{{% alert color="info" %}}
+Only a single character is supported. `character` parameters with more than one character may not work in all supported databases.
+{{% /alert %}}
+
+{{% alert color="info" %}}
+Like with other String functions, case sensitivity of the `LOCATE` function depends on the database. See [Behavior of Case Sensitivity by Database Type](/refguide/case-sensitive-database-behavior/#behavior-of-case-sensitivity-by-database-type) for details.
```
Comment (passalaqua) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: Same question as https://github.com/mendix/docs/pull/11058/changes#r3200000000

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 9]
```diff
@@ -1559,6 +1648,49 @@ ORDER BY LastName LIMIT 1
...
+##### character
+
+`character` is an optional single character string expression containing the character which will be removed from the beginning and end of the string. If omitted, the space character is used.
+
+{{% alert color="info" %}}
+Only a single character is supported. `character` parameters with more than one character may not work in all supported databases.
+{{% /alert %}}
+
+{{% alert color="info" %}}
+Like with other String functions, case sensitivity of the `LOCATE` function depends on the database. See [Behavior of Case Sensitivity by Database Type](/refguide/case-sensitive-database-behavior/#behavior-of-case-sensitivity-by-database-type) for details.
```
Comment (passalaqua) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: Same question as https://github.com/mendix/docs/pull/11058/changes#r3200000000

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 7 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: @passalaqua 
Forgot to update the expression function name but, basically, I wondered if there is a case sensitivity issue here. If the database is case insensitive would LTRIM(LastName. 'D') and LTRIM(LastName, 'd') do the same thing? Or does case sensitivity not apply to TRIM?

(GPT seems to think that TRIM is always case sensitive, so I guess I can remove this).

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 6 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: I'll leave this then - it seems that it is too much of an edge case and down to the customer to test.

**Review by MarkvanMents — APPROVED** [review]
Thanks for the PR and for working with me on the review.
Looks good.
Approving ready for publication with 11.11.
